core/src/main/java/org/kiwix/kiwixmobile/core/main/TabsAdapter.java

```diff
@@ -178,12 +186,42 @@ void setTabClickListener(TabClickListener listener) {
     final ImageView content;
     final TextView title;
     final ImageView close;
+    private NightModeConfig nightModeConfig;
+    private final Paint invertedPaint = createInvertedPaint();
 
-    ViewHolder(View v, ImageView content, TextView title, ImageView close) {
+    ViewHolder(View v, ImageView content, TextView title, ImageView close, Context context) {
       super(v);
       this.content = content;
       this.title = title;
       this.close = close;
+
+      SharedPreferenceUtil sharedPreferenceUtil = new SharedPreferenceUtil(context);
+      nightModeConfig = new NightModeConfig(sharedPreferenceUtil, context);
+
+      updateNightMode();
+    }
+
+    @NotNull private Paint createInvertedPaint() {
+      Paint paint = new Paint();
+      ColorMatrixColorFilter filterInvert = new ColorMatrixColorFilter(NIGHT_MODE_COLORS);
+      paint.setColorFilter(filterInvert);
+      return paint;
+    }
+
+    private void updateNightMode() {
```

Comment:
This entire feature is now duplicated across 2 classes. Create a class to encapsulate this, some like `NightModeViewPainter`, it should have one method `update(view)`

Reply:
I'll do the same for the createInvertedPaint() method too.

Reply:
@macgills the functionality of methods in both classes is same, but the methods differ. 
In KiwixWebView, the the methods are:
```java
public void deactivateNightMode() {
    setLayerType(LAYER_TYPE_NONE, null);
    videoView.setLayerType(LAYER_TYPE_NONE, null);
  }

  public void activateNightMode() {
    if (getUrl() != null && getUrl().equals(HOME_URL)) {
      return;
    }
    setLayerType(LAYER_TYPE_HARDWARE, invertedPaint);
    videoView.setLayerType(LAYER_TYPE_HARDWARE, invertedPaint);
  }
```

How do we keep these methods the same if we encapsulate the class?

Reply:
Should I create more than just the `update(view)` method and overload the methods?

Reply:
Hmm a bit annoying, kotlin gives us a way of working with this but it isn't really pretty

something like 

`fun updateView<T:View>(view:T, shouldActivateCriteria: T.()->Boolean=null, vararg additionalViews:View)`

I freehanded that so the syntax could be off. Touches on a few topics that could be a bit complex.

Try it if you want, otherwise I can live with only 2 duplications

Reply:
@yash1ts did you have trouble with this `updateView<T:View>(view:T, shouldActivateCriteria: T.()->Boolean=null, vararg additionalViews:View)` signature?

I can go into more depth if you need help as I am back on the project for this week and am not just doing drive by reviews.

Reply:
@macgills I'll try using it again by myself once. if it doesn't work I'll definitely ask for more clarification. Thanks for the guidance!

Reply:
@macgills I modifed the NightModeViewPainter like below. Is it fine:

```kotlin
class NightModeViewPainter @Inject constructor(
  private val nightModeConfig: NightModeConfig
) {

  private val invertedPaint =
    Paint().apply { colorFilter = ColorMatrixColorFilter(KiwixWebView.NIGHT_MODE_COLORS) }

  fun update(vararg additionalViews: View) {
    if (nightModeConfig.isNightModeActive()) {
      for (view in additionalViews) {
        activateNightMode(view)
      }
    } else {
      for (view in additionalViews) {
        deactivateNightMode(view)
      }
    }
  }

  fun deactivateNightMode(view: View) {
    view.setLayerType(View.LAYER_TYPE_NONE, null)
  }

  private fun activateNightMode(view: View) {
    view.setLayerType(View.LAYER_TYPE_HARDWARE, invertedPaint)
  }
}
```

Reply:
`if (getUrl() != null && getUrl().equals(HOME_URL)) {` what about this though?

Reply:
I put that condition when I'm calling update from CoreMainActivity.
```java
 private void updateNightMode() {
    if (getCurrentWebView().getUrl() != null && getCurrentWebView().getUrl().equals(HOME_URL)) {
      return;
    }

    painter.update(getCurrentWebView(), videoView);
  }
```

Reply:
This will change the logic, originally it is only triggered for activation not for deactivation.

This checks the condition for both cases.

That was the point of `shouldActivateCriteria: T.()->Boolean=null` which as a [receiver function](https://medium.com/tompee/idiomatic-kotlin-lambdas-with-receiver-and-dsl-3cd3348e1235) typed off of our first argument would have given us access to that arguments methods to perform the check.

Reply:
> originally it is only triggered for activation not for deactivation.

it is still triggered only in case nightMode is on right?
I modified the condition and included `nightModeConfig.isNightModeActive()` so that it only returns when night mode is on.

Reply:
Ah sorry, I didn't see the full width of that if statement.

I still think it is more confusing to do it this way. 
"if the url is not null and is home and night mode is active then don't update" as an expression is hiding the intention of the code
"don't activate night mode on the home screen"

Reply:
> That was the point of shouldActivateCriteria: T.()->Boolean=null which as a receiver function typed off of our first argument would have given us access to that arguments methods to perform the check.

I'll try to go ahead with this approach then.

Reply:
@macgills I'm facing some difficulties with this approach. 
```kotlin
fun <T : View> update(
    view: T,
    shouldActivateCriteria: (T.() -> Boolean)? = null,
    vararg additionalViews: View
  ) {
    if (shouldActivateCriteria != null) {
      if (view.shouldActivateCriteria()) {
        //call activateNightMode
      }
    }
  }
```

Is this correct?
And now how do I call this function from CoreMainActivity and TabsAdapter?

Reply:
sorry mislead you a little bit

`shouldActivateCriteria: (T.() -> Boolean)? = null,`

should be

`shouldActivateCriteria: ((T) -> Boolean) = { true },`

so this parameter is "a function which accepts a T and returns a Boolean, the default value of which is a function that returns true"

in kotlin usage would look like

```
painter.update(webView, { it.url == null || it.url !=HOME_URL}, videoView)
```

Reply:
@macgills I tried this. I haven't created the body of the `update` function yet, but I'm still unable to call the method. Could you please tell me where I'm going wrong.